### PR TITLE
[EOC-399 ] Refactor locked clients map

### PR DESCRIPTION
**Pull request by alemikolo**

### CODE REVIEW

File: client/src/modules/list/components/Items/ListItem/ListItem.jsx [hunk 1]
```diff
@@ -149,9 +145,11 @@ class ListItem extends PureComponent {
       }
     } = this.props;
 
-    this.lockItem();
+    this.handleItemLock();
 
-    return archiveItem(listId, itemId, name);
+    return archiveItem(listId, itemId, name).finally(() =>
```
Comment (adamklepacz) [inline client/src/modules/list/components/Items/ListItem/ListItem.jsx]: finally(this.handleItemUnlock)

File: client/src/modules/list/components/Items/ListItemDescription/ListItemDescription.jsx [hunk 2]
```diff
@@ -85,7 +80,6 @@ class ListItemDescription extends PureComponent {
 
   handleFocus = () => {
     const { onFocus } = this.props;
-
```
Comment (adamklepacz) [inline client/src/modules/list/components/Items/ListItemDescription/ListItemDescription.jsx]: why this change

File: server/sockets/cohort.js [hunk 3]
```diff
@@ -134,52 +136,52 @@ const updateCohort = (socket, allCohortsViewClients) => {
...
-
-    cohortClientLocks.set(userId, delayedUnlock);
+        .to(cohortChannel(cohortId))
+        .emit(CohortHeaderStatusTypes.LOCK, { cohortId, locks });
+
+      const lock = isDefined(nameLock) ? lockName : lockDescription;
+      const delayedUnlock = setTimeout(() => {
+        if (isDefined(nameLock)) {
+          locks.name = false;
+        } else {
```
Comment (adamklepacz) [inline server/sockets/cohort.js]: unnecessary else

File: server/sockets/list.js [hunk 4]
```diff
@@ -73,47 +75,87 @@ const restoreItem = socket => {
...
+          {
+            _id: listId,
+            'items._id': itemId,
+            memberIds: userId
+          },
+          itemId
+        )(locks).then(() => {
+          socket.broadcast
+            .to(listChannel(listId))
+            .emit(ItemStatusType.UNLOCK, { itemId, listId, locks });
```
Comment (adamklepacz) [inline server/sockets/list.js]: line of a break

File: server/sockets/list.js [hunk 5]
```diff
@@ -73,47 +75,87 @@ const restoreItem = socket => {
...
         .to(listChannel(listId))
-        .emit(ItemStatusType.UNLOCK, updatedData);
+        .emit(ItemStatusType.LOCK, { itemId, listId, locks });
+    });
 
-      itemClientLocks.delete(userId);
-    }, LOCK_TIMEOUT);
+    if (isDefined(nameLock)) {
+      const delayedUnlock = setTimeout(() => {
+        locks.name = false;
```
Comment (adamklepacz) [inline server/sockets/list.js]: line of a break

File: server/sockets/list.js [hunk 6]
```diff
@@ -73,47 +75,87 @@ const restoreItem = socket => {
...
+      }, LOCK_TIMEOUT);
 
-    itemClientLocks.set(userId, delayedUnlock);
+      itemClientLocks.set(lockName(itemId), delayedUnlock);
+    }
 
-    const locks = { description: descriptionLock, name: nameLock };
+    if (isDefined(descriptionLock)) {
+      const delayedUnlock = setTimeout(() => {
+        locks.description = false;
```
Comment (adamklepacz) [inline server/sockets/list.js]: line o a break

File: server/sockets/list.js [hunk 7]
```diff
@@ -73,47 +75,87 @@ const restoreItem = socket => {
...
+          {
+            _id: listId,
+            'items._id': itemId,
+            memberIds: userId
+          },
+          { itemId }
+        )(locks).then(() => {
+          socket.broadcast
+            .to(listChannel(listId))
+            .emit(ItemStatusType.UNLOCK, { itemId, listId, locks });
```
Comment (adamklepacz) [inline server/sockets/list.js]: line of a break

File: server/sockets/list.js [hunk 8]
```diff
@@ -73,47 +75,87 @@ const restoreItem = socket => {
...
+      }, LOCK_TIMEOUT);
 
-    handleItemLocks(List, { _id: listId, 'items._id': itemId }, itemId)(locks);
+      itemClientLocks.set(lockName(itemId), delayedUnlock);
+    }
   });
 
   socket.on(ItemStatusType.UNLOCK, data => {
     const { descriptionLock, itemId, listId, nameLock, userId } = data;
+    const locks = { description: descriptionLock, name: nameLock };
```
Comment (adamklepacz) [inline server/sockets/list.js]: line o a break

File: server/sockets/list.js [hunk 9]
```diff
@@ -305,52 +347,49 @@ const updateList = (socket, dashboardViewClients, cohortViewClients) => {
...
-
-      if (isDefined(nameLock)) {
-        updatedData.nameLock = false;
-      }
-
-      if (isDefined(descriptionLock)) {
-        updatedData.descriptionLock = false;
-      }
-
+    const locks = { description: descriptionLock, name: nameLock };
```
Comment (adamklepacz) [inline server/sockets/list.js]: line o a break

File: server/sockets/list.js [hunk 10]
```diff
@@ -305,52 +347,49 @@ const updateList = (socket, dashboardViewClients, cohortViewClients) => {
...
-
-    listClientLocks.set(userId, delayedUnlock);
+        .to(listChannel(listId))
+        .emit(ListHeaderStatusTypes.LOCK, { listId, locks });
+
+      const lock = isDefined(nameLock) ? lockName : lockDescription;
+      const delayedUnlock = setTimeout(() => {
+        if (isDefined(nameLock)) {
+          locks.name = false;
+        } else {
```
Comment (adamklepacz) [inline server/sockets/list.js]: unnecessary else

File: server/sockets/list.js [hunk 11]
```diff
@@ -305,52 +347,49 @@ const updateList = (socket, dashboardViewClients, cohortViewClients) => {
...
+          locks.description = false;
+        }
 
-    const locks = { description: descriptionLock, name: nameLock };
+        handleLocks(List, { _id: listId, ownerIds: userId })(locks).then(() => {
+          socket.broadcast
+            .to(listChannel(listId))
+            .emit(ListHeaderStatusTypes.UNLOCK, { listId, locks });
```
Comment (adamklepacz) [inline server/sockets/list.js]: line o a break

File: server/sockets/helpers/index.js [hunk 12]
```diff
@@ -186,6 +186,10 @@ const handleItemLocks = (model, query, itemId) => ({ description, name }) =>
       }
     });
 
+const lockName = cohortId => `name-${cohortId}`;
```
Comment (Ciunkos) [inline server/sockets/helpers/index.js]: I would go for `nameLockId` and `descriptionLockId`.

File: client/src/modules/list/components/Items/ListItemDescription/ListItemDescription.jsx (hunk 2 above)
Comment (alemikolo) [reply client/src/modules/list/components/Items/ListItemDescription/ListItemDescription.jsx]: onFocus and onBlur were invoked twice and locks were set to true (or false) twice because socket events were emitting twice; I had problems with these behaviours so I removed those lines and realized that app is still fine but events aren't emitting twice...

File: server/sockets/cohort.js (hunk 3 above)
Comment (alemikolo) [reply server/sockets/cohort.js]: it is necessary, without else description lock will always set to false. In lists and cohort case one user can set name OR description lock so I had to use else. I can use:
```
if (isDefined(nameLock)) {
          locks.name = false;
        } 
if (isDefined(descriptionLock)) {
          locks.description = false;
        } 
```
but in this case we would have unnecessary check

File: server/sockets/list.js (hunk 10 above)
Comment (alemikolo) [reply server/sockets/list.js]: see answer above

File: client/src/modules/list/components/Items/ListItemDescription/ListItemDescription.jsx (hunk 2 above)
Comment (adamklepacz) [reply client/src/modules/list/components/Items/ListItemDescription/ListItemDescription.jsx]: no no, I meant - why did you remove that empty line :D

File: server/sockets/cohort.js (hunk 3 above)
Comment (adamklepacz) [reply server/sockets/cohort.js]: I think this is better and more reasonable:
```
if (isDefined(nameLock)) {
          locks.name = false;
        } 
if (isDefined(descriptionLock)) {
          locks.description = false;
        } 
```

File: client/src/modules/list/components/Items/ListItemDescription/ListItemDescription.jsx (hunk 2 above)
Comment (alemikolo) [reply client/src/modules/list/components/Items/ListItemDescription/ListItemDescription.jsx]: ok, I added that line
